pages/shared/editHelpers.js

```diff
@@ -0,0 +1,680 @@
+                        job.fail(new Error(`Another user is already syncing or modifying the course: ${options.courseDir}`));
+                    } else {
+                        courseLock = lock;
+                        job.verbose(`Acquired lock ${lockName}`);
+                        job.succeed();
+                    }
+                    return;
+                });
+            });
+        };
+
+        const _clean = (on_success, on_failure) => {
+            debug(`${job_sequence_id}: _clean`);
+            const jobOptions = {
+                course_id: options.course_id,
+                user_id: options.user_id,
+                authn_user_id: options.authn_user_id,
+                job_sequence_id: job_sequence_id,
+                type: 'clean_git_repo',
+                description: 'Clean local files not in remote git repository',
+                command: 'git',
+                arguments: ['clean', '-fdx'],
+                working_directory: edit.coursePath,
+                env: gitEnv,
+                on_success: () => {_reset(on_success, on_failure);},
+                on_error: on_failure,
+                no_job_sequence_update: true,
+            };
+            serverJobs.spawnJob(jobOptions);
+        };
+
+        const _reset = (on_success, on_failure) => {
+            debug(`${job_sequence_id}: _reset`);
+            const jobOptions = {
+                course_id: options.course_id,
+                user_id: options.user_id,
+                authn_user_id: options.authn_user_id,
+                job_sequence_id: job_sequence_id,
+                type: 'reset_from_git',
+                description: 'Reset state to remote git repository',
+                command: 'git',
+                arguments: ['reset', '--hard', 'origin/master'],
+                working_directory: edit.coursePath,
+                env: gitEnv,
+                on_success: on_success,
+                on_error: on_failure,
+                no_job_sequence_update: true,
+            };
+            serverJobs.spawnJob(jobOptions);
+        };
+
+        const _write = () => {
+            debug(`${job_sequence_id}: _write`);
+            const jobOptions = {
+                course_id: options.course_id,
+                user_id: options.user_id,
+                authn_user_id: options.authn_user_id,
+                type: 'write',
+                description: 'Write to disk',
+                job_sequence_id: job_sequence_id,
+                on_success: (config.fileEditorUseGit ? _add : _unlock),
+                on_error: (config.fileEditorUseGit ? _cleanupAfterWrite : _cleanup),
+                no_job_sequence_update: true,
+            };
+            serverJobs.createJob(jobOptions, (err, job) => {
+                if (ERR(err, (e) => logger.error(e))) {
+                    _finishWithFailure();
+                    return;
+                }
+
+                edit.write(edit, (err) => {
```

Comment:
Would it be cleaner to use JS classes here and make an `Editor` class that contains its own data about the edit? Then we'd have a `RenameEditor`, `DeleteEditor`, etc? The current system is doing almost exactly this, but without using JS classes.

Reply:
Good idea. Done.